### ISSUE TITLE
"cudart64_101.dll not found" but the nvidia GPU computing toolkit has the file cudart64_102.dll file under the bin folder?

### ISSUE THREAD

**Issue by NahianAlindo** [tensorflow/tensorflow#42811]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10 Pro
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): using pip after creating environment from within the scripts folder of the environment after using "activate", so in short using binary
- TensorFlow version: tensorflow gpu 2.3.0
- Python version: 3.6
- Installed using virtualenv? pip? conda?: pip after creating an environment
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.2
- GPU model and memory: Nvidia GeForce 940MX

I tried to install tensorflow gpu to get started with using CUDA with it. But after installing it in the created environment, when I open the python interactive mode and try to import tensorflow it throws a error/warning that cudart64_101.dll not found. But installed cuda 10.2 and then it's patch listed on the nvidia cuda website and also cudnn. I even included everything in the PATH. But why is my installation looking for cudart64_101.dll when I have an updated file named cudart64_102.dll in my nvidia gpu toolkit's binary folder. How can I fix this? I am trying this to learn about cuda and how to use them with tensorflow and pytorch both. The commands I used from start to finish and their output is pasted below:

**Provide the exact sequence of commands / steps that you executed before running into the problem**
C:\Users\nrifa>cd C:\Users\nrifa\AppData\Local\Programs\Python\Python36

C:\Users\nrifa\AppData\Local\Programs\Python\Python36>python -m venv C:\gputest

C:\Users\nrifa\AppData\Local\Programs\Python\Python36>cd C:\gputest

C:\gputest>cd Scripts

C:\gputest\Scripts>activate
(gputest) C:\gputest\Scripts>pip install tensorflow-gpu

Logs:

Collecting tensorflow-gpu
  Downloading https://files.pythonhosted.org/packages/9f/23/5d5a18870f4be65faee538aeb28276af809f6442a3a5dc6f7e58d7f70bd6/tensorflow_gpu-2.3.0-cp36-cp36m-win_amd64.whl (344.1MB)
    100% |████████████████████████████████| 344.1MB 3.2kB/s
Collecting keras-preprocessing<1.2,>=1.1.1 (from tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/79/4c/7c3275a01e12ef9368a892926ab932b33bb13d55794881e3573482b378a7/Keras_Preprocessing-1.1.2-py2.py3-none-any.whl (42kB)
    100% |████████████████████████████████| 51kB 631kB/s
Collecting h5py<2.11.0,>=2.10.0 (from tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/0b/fa/bee65d2dbdbd3611702aafd128139c53c90a1285f169ba5467aab252e27a/h5py-2.10.0-cp36-cp36m-win_amd64.whl (2.4MB)
    100% |████████████████████████████████| 2.4MB 321kB/s
Collecting termcolor>=1.1.0 (from tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/8a/48/a76be51647d0eb9f10e2a4511bf3ffb8cc1e6b14e9e4fab46173aa79f981/termcolor-1.1.0.tar.gz
Collecting tensorboard<3,>=2.3.0 (from tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/e9/1b/6a420d7e6ba431cf3d51b2a5bfa06a958c4141e3189385963dc7f6fbffb6/tensorboard-2.3.0-py3-none-any.whl (6.8MB)
    100% |████████████████████████████████| 6.8MB 133kB/s
Collecting wheel>=0.26 (from tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/a7/00/3df031b3ecd5444d572141321537080b40c1c25e1caa3d86cdd12e5e919c/wheel-0.35.1-py2.py3-none-any.whl
Collecting scipy==1.4.1 (from tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/8d/2f/fcb6150813b89d628749784370132e431f687ebab5a1063eb298cc941f76/scipy-1.4.1-cp36-cp36m-win_amd64.whl (30.8MB)
    100% |████████████████████████████████| 30.8MB 16kB/s
Collecting gast==0.3.3 (from tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/d6/84/759f5dd23fec8ba71952d97bcc7e2c9d7d63bdc582421f3cd4be845f0c98/gast-0.3.3-py2.py3-none-any.whl
Collecting absl-py>=0.7.0 (from tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/b9/07/f69dd3367368ad69f174bfe426a973651412ec11d48ec05c000f19fe0561/absl_py-0.10.0-py3-none-any.whl (127kB)
    100% |████████████████████████████████| 133kB 141kB/s
Collecting protobuf>=3.9.2 (from tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/f6/fe/9d8e70a86add02cb1ef35540ec03fd5b210d76323fe4645d7121b13ae33e/protobuf-3.13.0-cp36-cp36m-win_amd64.whl (1.1MB)
    100% |████████████████████████████████| 1.1MB 271kB/s
Collecting numpy<1.19.0,>=1.16.0 (from tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/dc/18/e69ef84530360c2d39db51acb4cc0012990f27fb1fa7542dac45ad30e7ab/numpy-1.18.5-cp36-cp36m-win_amd64.whl (12.7MB)
    100% |████████████████████████████████| 12.7MB 34kB/s
Collecting tensorflow-gpu-estimator<2.4.0,>=2.3.0 (from tensorflow-gpu)
  Downloading https://files.pythonhosted.org/packages/9d/6f/87e922b1dbfa9aa2e79bf150bf05d567eaa4f83bfd329c04834b26b0725e/tensorflow_gpu_estimator-2.3.0-py2.py3-none-any.whl (474kB)
    100% |████████████████████████████████| 481kB 266kB/s
Collecting astunparse==1.6.3 (from tensorflow-gpu)
  Downloading https://files.pythonhosted.org/packages/2b/03/13dde6512ad7b4557eb792fbcf0c653af6076b81e5941d36ec61f7ce6028/astunparse-1.6.3-py2.py3-none-any.whl
Collecting wrapt>=1.11.1 (from tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/82/f7/e43cefbe88c5fd371f4cf0cf5eb3feccd07515af9fd6cf7dbf1d1793a797/wrapt-1.12.1.tar.gz
Collecting six>=1.12.0 (from tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/ee/ff/48bde5c0f013094d729fe4b0316ba2a24774b3ff1c52d924a8a4cb04078a/six-1.15.0-py2.py3-none-any.whl
Collecting grpcio>=1.8.6 (from tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/15/3f/f311f382bb658387fe78a30e1ed55193fe94c5e78b37abd134c34bd256eb/grpcio-1.31.0-cp36-cp36m-win_amd64.whl (2.6MB)
    100% |████████████████████████████████| 2.6MB 269kB/s
Collecting google-pasta>=0.1.8 (from tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/a3/de/c648ef6835192e6e2cc03f40b19eeda4382c49b5bafb43d88b931c4c74ac/google_pasta-0.2.0-py3-none-any.whl (57kB)
    100% |████████████████████████████████| 61kB 1.2MB/s
Collecting opt-einsum>=2.3.2 (from tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/bc/19/404708a7e54ad2798907210462fd950c3442ea51acc8790f3da48d2bee8b/opt_einsum-3.3.0-py3-none-any.whl (65kB)
    100% |████████████████████████████████| 71kB 997kB/s
Collecting markdown>=2.6.8 (from tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/a4/63/eaec2bd025ab48c754b55e8819af0f6a69e2b1e187611dd40cbbe101ee7f/Markdown-3.2.2-py3-none-any.whl (88kB)
    100% |████████████████████████████████| 92kB 963kB/s
Collecting werkzeug>=0.11.15 (from tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/cc/94/5f7079a0e00bd6863ef8f1da638721e9da21e5bacee597595b318f71d62e/Werkzeug-1.0.1-py2.py3-none-any.whl (298kB)
    100% |████████████████████████████████| 307kB 363kB/s
Collecting tensorboard-plugin-wit>=1.6.0 (from tensorboard<3,>=2.3.0->tensorflow-gpu)
  Downloading https://files.pythonhosted.org/packages/b6/85/5c5ac0a8c5efdfab916e9c6bc18963f6a6996a8a1e19ec4ad8c9ac9c623c/tensorboard_plugin_wit-1.7.0-py3-none-any.whl (779kB)
    100% |████████████████████████████████| 788kB 423kB/s
Collecting setuptools>=41.0.0 (from tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/c3/a9/5dc32465951cf4812e9e93b4ad2d314893c2fa6d5f66ce5c057af6e76d85/setuptools-49.6.0-py3-none-any.whl (803kB)
    100% |████████████████████████████████| 808kB 248kB/s
Collecting google-auth<2,>=1.6.3 (from tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/63/7f/ef6bcf2cc0f50c7163afb94382aab67a6b278e1e447c2e3981aa281b9747/google_auth-1.21.0-py2.py3-none-any.whl (92kB)
    100% |████████████████████████████████| 102kB 366kB/s
Collecting google-auth-oauthlib<0.5,>=0.4.1 (from tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/7b/b8/88def36e74bee9fce511c9519571f4e485e890093ab7442284f4ffaef60b/google_auth_oauthlib-0.4.1-py2.py3-none-any.whl
Collecting requests<3,>=2.21.0 (from tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/45/1e/0c169c6a5381e241ba7404532c16a21d86ab872c9bed8bdcd4c423954103/requests-2.24.0-py2.py3-none-any.whl (61kB)
    100% |████████████████████████████████| 71kB 432kB/s
Collecting importlib-metadata; python_version < "3.8" (from markdown>=2.6.8->tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/8e/58/cdea07eb51fc2b906db0968a94700866fc46249bdc75cac23f9d13168929/importlib_metadata-1.7.0-py2.py3-none-any.whl
Collecting pyasn1-modules>=0.2.1 (from google-auth<2,>=1.6.3->tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/95/de/214830a981892a3e286c3794f41ae67a4495df1108c3da8a9f62159b9a9d/pyasn1_modules-0.2.8-py2.py3-none-any.whl (155kB)
    100% |████████████████████████████████| 163kB 475kB/s
Collecting cachetools<5.0,>=2.0.0 (from google-auth<2,>=1.6.3->tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/cd/5c/f3aa86b6d5482f3051b433c7616668a9b96fbe49a622210e2c9781938a5c/cachetools-4.1.1-py3-none-any.whl
Collecting rsa<5,>=3.1.4; python_version >= "3.5" (from google-auth<2,>=1.6.3->tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/1c/df/c3587a667d6b308fadc90b99e8bc8774788d033efcc70f4ecaae7fad144b/rsa-4.6-py3-none-any.whl (47kB)
    100% |████████████████████████████████| 51kB 506kB/s
Collecting requests-oauthlib>=0.7.0 (from google-auth-oauthlib<0.5,>=0.4.1->tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/a3/12/b92740d845ab62ea4edf04d2f4164d82532b5a0b03836d4d4e71c6f3d379/requests_oauthlib-1.3.0-py2.py3-none-any.whl
Collecting idna<3,>=2.5 (from requests<3,>=2.21.0->tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/a2/38/928ddce2273eaa564f6f50de919327bf3a00f091b5baba8dfa9460f3a8a8/idna-2.10-py2.py3-none-any.whl (58kB)
    100% |████████████████████████████████| 61kB 501kB/s
Collecting certifi>=2017.4.17 (from requests<3,>=2.21.0->tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/5e/c4/6c4fe722df5343c33226f0b4e0bb042e4dc13483228b4718baf286f86d87/certifi-2020.6.20-py2.py3-none-any.whl (156kB)
    100% |████████████████████████████████| 163kB 312kB/s
Collecting urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 (from requests<3,>=2.21.0->tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/9f/f0/a391d1463ebb1b233795cabfc0ef38d3db4442339de68f847026199e69d7/urllib3-1.25.10-py2.py3-none-any.whl (127kB)
    100% |████████████████████████████████| 133kB 412kB/s
Collecting chardet<4,>=3.0.2 (from requests<3,>=2.21.0->tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/bc/a9/01ffebfb562e4274b6487b4bb1ddec7ca55ec7510b22e4c51f14098443b8/chardet-3.0.4-py2.py3-none-any.whl (133kB)
    100% |████████████████████████████████| 143kB 506kB/s
Collecting zipp>=0.5 (from importlib-metadata; python_version < "3.8"->markdown>=2.6.8->tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/b2/34/bfcb43cc0ba81f527bc4f40ef41ba2ff4080e047acb0586b56b3d017ace4/zipp-3.1.0-py3-none-any.whl
Collecting pyasn1<0.5.0,>=0.4.6 (from pyasn1-modules>=0.2.1->google-auth<2,>=1.6.3->tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/62/1e/a94a8d635fa3ce4cfc7f506003548d0a2447ae76fd5ca53932970fe3053f/pyasn1-0.4.8-py2.py3-none-any.whl (77kB)
    100% |████████████████████████████████| 81kB 101kB/s
Collecting oauthlib>=3.0.0 (from requests-oauthlib>=0.7.0->google-auth-oauthlib<0.5,>=0.4.1->tensorboard<3,>=2.3.0->tensorflow-gpu)
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/05/57/ce2e7a8fa7c0afb54a0581b14a65b56e62b5759dbc98e80627142b8a3704/oauthlib-3.1.0-py2.py3-none-any.whl (147kB)
    100% |████████████████████████████████| 153kB 254kB/s
Installing collected packages: numpy, six, keras-preprocessing, h5py, termcolor, wheel, zipp, importlib-metadata, markdown, grpcio, werkzeug, setuptools, protobuf, tensorboard-plugin-wit, pyasn1, pyasn1-modules, cachetools, rsa, google-auth, idna, certifi, urllib3, chardet, requests, oauthlib, requests-oauthlib, google-auth-oauthlib, absl-py, tensorboard, scipy, gast, tensorflow-gpu-estimator, astunparse, wrapt, google-pasta, opt-einsum, tensorflow-gpu
  Running setup.py install for termcolor ... done
  Found existing installation: setuptools 28.8.0
    Uninstalling setuptools-28.8.0:
      Successfully uninstalled setuptools-28.8.0
  Running setup.py install for wrapt ... done
Successfully installed absl-py-0.10.0 astunparse-1.6.3 cachetools-4.1.1 certifi-2020.6.20 chardet-3.0.4 gast-0.3.3 google-auth-1.21.0 google-auth-oauthlib-0.4.1 google-pasta-0.2.0 grpcio-1.31.0 h5py-2.10.0 idna-2.10 importlib-metadata-1.7.0 keras-preprocessing-1.1.2 markdown-3.2.2 numpy-1.18.5 oauthlib-3.1.0 opt-einsum-3.3.0 protobuf-3.13.0 pyasn1-0.4.8 pyasn1-modules-0.2.8 requests-2.24.0 requests-oauthlib-1.3.0 rsa-4.6 scipy-1.4.1 setuptools-49.6.0 six-1.15.0 tensorboard-2.3.0 tensorboard-plugin-wit-1.7.0 tensorflow-gpu-2.3.0 tensorflow-gpu-estimator-2.3.0 termcolor-1.1.0 urllib3-1.25.10 werkzeug-1.0.1 wheel-0.35.1 wrapt-1.12.1 zipp-3.1.0
You are using pip version 9.0.1, however version 20.2.2 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.

(gputest) C:\gputest\Scripts>python -m pip install --upgrade pip
Cache entry deserialization failed, entry ignored
Collecting pip
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/5a/4a/39400ff9b36e719bdf8f31c99fe1fa7842a42fa77432e584f707a5080063/pip-20.2.2-py2.py3-none-any.whl (1.5MB)
    100% |████████████████████████████████| 1.5MB 230kB/s
Installing collected packages: pip
  Found existing installation: pip 9.0.1
    Uninstalling pip-9.0.1:
      Successfully uninstalled pip-9.0.1
Successfully installed pip-20.2.2

(gputest) C:\gputest\Scripts>python
Python 3.6.4 (v3.6.4:d48eceb, Dec 19 2017, 06:54:40) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow
2020-08-30 21:51:17.099536: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'cudart64_101.dll'; dlerror: cudart64_101.dll not found
2020-08-30 21:51:17.100010: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.

**Comment by Saduf2019** [comment]
@NahianAlindo 
You may need to find which directory these dlls are in, and then update your PATH to include that directory. 
Can you refer to [this](https://stackoverflow.com/questions/59823283/could-not-load-dynamic-library-cudart64-101-dll-on-tensorflow-cpu-only-install) issue [#36111] with same error and let us know if it helps.

**Comment by NahianAlindo** [comment]
> @NahianAlindo
> You may need to find which directory these dlls are in, and then update your PATH to include that directory.
> Can you refer to [this](https://stackoverflow.com/questions/59823283/could-not-load-dynamic-library-cudart64-101-dll-on-tensorflow-cpu-only-install) issue [#36111] with same error and let us know if it helps.

I think I got where the issue was reading the tensorflow website. Apparently, tensorflow 2.3.0 which is the latest version of tensorflow requires cuda 10.1 but I  have installed cuda 10.2. I will try uninstalling and reinstalling. Maybe that will fix things up.

**Comment by Smankusors** [comment]
on my Windows, I just do symbolic link and the tensorflow just runs fine on me 😉

**Comment by Saduf2019** [comment]
@NahianAlindo
Can you please update.

**Comment by NahianAlindo** [comment]
> @NahianAlindo
> Can you please update.

Yes, I have fixed it just now. It seems that the TensorFlow 2.3.0 requires CUDA 10.1 after all. That's why it was searching for cudart64_101.dll instead of what I had which was cudart64_102.dll. So, I uninstalled the CUDA toolkit 10.2 and installed the CUDA toolkit 10.1 and the latest CUDNN for the CUDA toolkit 10.1. Thus, I managed to make GPU available for both TensorFlow and PyTorch.
![2](https://user-images.githubusercontent.com/37206331/92404159-b685f280-f154-11ea-9bef-08165d0460a8.PNG)
